testsuite/forge/src/tx_emitter.rs

```diff
@@ -0,0 +1,717 @@
+                        // Simplified from:
+                        // end_time * num_committed - (tx_offset_time/num_requests) * num_committed
+                        // to
+                        // (end_time - tx_offset_time / num_requests) * num_committed
+                        latency * num_committed as u64,
+                        Ordering::Relaxed,
+                    );
+                    self.stats
+                        .latencies
+                        .record_data_point(latency, num_committed);
+                    info!(
+                        "[{:?}] Transactions were not committed before expiration: {:?}",
+                        self.client, uncommitted
+                    );
+                } else {
+                    let end_time = (Instant::now() - start_time).as_millis() as u64;
+                    let latency = end_time - tx_offset_time / num_requests as u64;
+                    self.stats
+                        .committed
+                        .fetch_add(num_requests as u64, Ordering::Relaxed);
+                    self.stats
+                        .latency
+                        .fetch_add(latency * num_requests as u64, Ordering::Relaxed);
+                    self.stats
+                        .latencies
+                        .record_data_point(latency, num_requests as u64);
+                }
+            }
+            let now = Instant::now();
+            if wait_util > now {
+                time::sleep(wait_util - now).await;
+            }
+        }
+        self.accounts
+    }
+
+    fn gen_requests(&mut self) -> Vec<SignedTransaction> {
+        let mut rng = ThreadRng::default();
+        let batch_size = max(MAX_TXN_BATCH_SIZE, self.accounts.len());
+        let accounts = self
+            .accounts
+            .iter_mut()
+            .choose_multiple(&mut rng, batch_size);
+        let mut requests = Vec::with_capacity(accounts.len());
+        for sender in accounts {
+            let receiver = self
+                .all_addresses
+                .choose(&mut rng)
+                .expect("all_addresses can't be empty");
+            let request =
+                gen_transfer_txn_request(sender, receiver, SEND_AMOUNT, self.tx_factory.clone());
+            requests.push(request);
+        }
+        requests
+    }
+}
+
+#[derive(Debug)]
+pub struct TxEmitter<'t> {
+    accounts: Vec<LocalAccount>,
+    tx_factory: TransactionFactory,
+    network_info: NetworkInfo<'t>,
+    client: JsonRpcClient,
+}
+
+impl<'t> TxEmitter<'t> {
+    pub fn new(network_info: NetworkInfo<'t>) -> Self {
+        let tx_factory = TransactionFactory::new(network_info.chain_id());
+        let client = JsonRpcClient::new(network_info.json_rpc());
+        Self {
+            accounts: vec![],
+            tx_factory,
+            network_info,
+            client,
+        }
+    }
+
+    pub async fn get_money_source(&mut self, coins_total: u64) -> Result<&mut LocalAccount> {
```

Comment:
nit: We can probably drop the `get_` and just go with `money_source`

Also, is there a reason for it to be `pub` if it's only used here?  Seems pretty self contained to the `TxEmitter`